[Feature:TAGrading] Add Image Annotation

martig7 opened the pull request:
### Why is this Change Important & Necessary?
PDF Annotation was recently removed, and so as a replacement we agreed to instead support image annotation.

### What is the New Behavior?


### What steps should a reviewer take to reproduce or test the bug or new feature?
Bulk upload to a gradable on the new branch, and assign the result to a student. Go to their submisson, and open up an image. This page should now show, and editing the image should render the changes. These changes should persist across instances. Downloading should download the image with the annotations built on top.
<img width="3839" height="2066" alt="image" src="https://github.com/user-attachments/assets/48bd5eac-5f9a-47e6-b4b7-b29e04e05621" />

The student can now see their annotated image from the gradable at the bottom of the page, which is the same as with pdf annotation:

<img width="3838" height="2233" alt="image" src="https://github.com/user-attachments/assets/a6ecfdec-cf3f-47c4-a985-bafbdf269a57" />
<img width="3813" height="2201" alt="image" src="https://github.com/user-attachments/assets/5bd13c94-a7c4-46f8-8b81-60bda1110266" />

### Automated Testing & Documentation
This feature will likely need submitty.org documentation.

### Other information
A future PR/issue should store this annotated image somewhere without having to build to see the annotated image.

martig7 committed:
Add markerjs package

martig7 committed:
Add image display with annotation buttons

martig7 committed:
add markerjs-ui

martig7 committed:
markerjs-ui attempt

martig7 committed:
Imported markerjs as a js instead of a cjs so that it can get seen

martig7 committed:
clean up image annotation frame

martig7 committed:
Refactor to use the same markerJS instance globally.

martig7 committed:
Broke some things, working towards fixing

martig7 committed:
Use MarkerView instead

martig7 committed:
Attempt to fix download issue, fix first render issue

martig7 committed:
Merge branch 'main' into add-image-annotation

Chriun reviewed (commented):
Clearing annotations does not clear the rendered view but it is cleared in the view when you click edit annotations. Downloading the image also contains the annotations that were supposed to be cleared.

martig7 committed:
Fix download without editor open

martig7 committed:
Refactor to use a class for state

martig7 committed:
Merge branch 'main' into add-image-annotation

martig7 committed:
port image annotation to typescript

martig7 committed:
Merge branch 'main' into add-image-annotation

martig7 committed:
First lint passover

martig7 committed:
Remove debug messages

martig7 committed:
Refactor addAnnotations to be a lot simpler looking, fix linting.

martig7 committed:
change requires to imports, make eslint config an mjs

martig7 committed:
Merge branch 'main' into add-image-annotation

martig7 committed:
Remove unused css, lint css

martig7 committed:
Fix lint, remove unused function

lavalleeale commented on site/app/views/ImageView.php:
```diff
@@ -0,0 +1,87 @@
+    private function getDirectoryFromPath(string $file_path): string {
+        if (str_contains($file_path, 'user_assignment_settings.json')) {
+            return 'submission_versions';
+        }
```
```suggestion
        }
        elseif (str_contains($file_path, 'submissions_processed')) {
            return 'submissions_processed';
        }
```

lavalleeale reviewed (changes requested):
The vertical stack of buttons does not look great on a small screen / multi panel layout
<img src="https://github.com/user-attachments/assets/e6aaf6ee-7ea5-4e80-925f-a6fddd8be916" />
In addition, most of the `console.log`s should be removed before this can get merged.
I also did not see the annotations I made as instructor when I logged in as ta.

martig7 committed:
Update site/app/views/ImageView.php

Co-authored-by: Alex Lavallee <alex@lavallee.one>

martig7 committed:
Remove console logs, annotation-data element, add prototype of multi-annotate view (commented out)

martig7 committed:
Improve pdf view

martig7 committed:
Fix bug I introduced, fix lint

martig7 committed:
Merge branch 'main' into add-image-annotation

martig7 committed:
Merge branch 'fix-windows-eslint' into add-image-annotation

martig7 committed:
Revert "Merge branch 'fix-windows-eslint' into add-image-annotation"

This reverts commit 09ea4ac2152c035bb43a1d7dbf944f4cab4fa307, reversing
changes made to 9c3720b5819ece81ccb534e929c43ce77a6e284b.

martig7 committed:
Fix Linting

Chriun reviewed (commented):
All image annotation functionality now works correctly on my end. Annotations don't seem to be shown in the preview but that isn't in the scope of this PR.

<img width="500" height="583" alt="image" src="https://github.com/user-attachments/assets/e2473a5e-f420-495f-b5e2-6cacd6f46349" />

Additionally, there are still functions that reference the old PDF annotator (savePDFAnnotation, showGraderPDFEmbedded) that aren't called anywhere in the codebase. These can be removed in another PR.

martig7 committed:
Add student view of annotated files

martig7 committed:
Add uncommited comment

martig7 committed:
Merge branch 'main' into add-image-annotation

martig7 committed:
Fix javascript lint... hopefuly...

martig7 committed:
More linting

martig7 committed:
Fix curly brace

Chriun reviewed (commented):
Student view of submissions now work.  

<img width="1452" height="188" alt="image" src="https://github.com/user-attachments/assets/e60bea08-72dd-4dc5-bc41-e527ed748352" />

But if an instructor clears their annotations, the student view still shows an option for popping out the image but it gives this error.

<img width="555" height="197" alt="image" src="https://github.com/user-attachments/assets/f10519c4-954c-4316-8625-bea6a0c6e820" />

martig7 committed:
Remove empty annotation files, check tryGet function returns

martig7 commented on site/app/controllers/pdf/PDFController.php:
```diff
@@ -77,6 +77,50 @@ public function showStudentPDF(string $gradeable_id, string $filename, string $p
+    #[Route("/courses/{_semester}/{_course}/gradeable/{gradeable_id}/img")]
+    public function showStudentImg(string $gradeable_id, string $filename, string $path, string $anon_path): JsonResponse {
+        $id = $this->core->getUser()->getId();
+        $gradeable = $this->tryGetGradeable($gradeable_id);
```
Fixed both, now annotation files with no markers will be deleted.

Chriun reviewed (commented):
Looks good from what I tested. Code may need a refactor and additional features (such as viewing annotations from other graders) in the future.

As of now, though, image annotation and student viewing of the graded annotated images works. Others should most likely test as well.

jeffrey-cordero reviewed (commented):
Overall, the image annotation features look very robust. I could only find an issue with being unable to view/download annotations on the student-view, which I'll look into after a VM reset, but otherwise, the annotation components worked well. Having a native save button within the modal itself, and clicking on `submissions_processed` automatically opening up the modal would be nice, but these can be future enhancements.

jeffrey-cordero commented on site/app/controllers/pdf/PDFController.php:
```diff
@@ -77,6 +77,53 @@ public function showStudentPDF(string $gradeable_id, string $filename, string $p
+            '?dir=submissions_processed' .
+            '&file=' . urlencode($filename) .
+            '&path=' . urlencode($path) .
+            '&ta_grading=true';
```
I've found a fix on my end. Add `'&view_annotated_file=true';` to this image URL, then update `MiscController.displayFile` method signature line `176` with the following lines.

```php
public function displayFile($dir = null, $path = null, $gradeable_id = null, $user_id = null, $ta_grading = null, $course_material_id = null, $view_annotated_file = false) {
...
// Check access through Access library
$owns_annotated_file = $view_annotated_file && $dir === 'submissions_processed' && strpos($path, $this->core->getUser()->getId()) !== false;
if (!$this->core->getAccess()->canI("path.read", ["dir" => $dir, "path" => $path]) && !$owns_annotated_file) {
    $this->core->getOutput()->showError(self::GENERIC_NO_ACCESS_MSG);
    return false;
}
```

martig7 replied:
Alright, I implemented your fix

jeffrey-cordero replied:
Looks good.

lavalleeale commented on site/app/controllers/pdf/PDFController.php:
```diff
@@ -128,7 +178,7 @@ public function downloadStudentPDF(string $gradeable_id, string $filename, strin
         }
 
-        $rerender_annotated_pdf = (file_exists($annotation_path) && $latest_timestamp <= filemtime($annotation_path)) !== true;
+        $rerender_annotated_pdf = !((file_exists($annotation_path) && $latest_timestamp <= filemtime($annotation_path)));
```
```suggestion
        $rerender_annotated_pdf = !(file_exists($annotation_path) && $latest_timestamp <= filemtime($annotation_path));
```
remove extra paren

lavalleeale commented on site/app/controllers/pdf/PDFController.php:
```diff
@@ -199,10 +249,10 @@ public function showGraderPDFEmbedded(string $gradeable_id) {
         $is_anon = $_POST['is_anon'] ?? false;
```
```suggestion
        $is_anon = (bool) $_POST['is_anon'] ?? false;
```
Let's do the conversion once instead of everywhere it's used (this applies to `showGraderImageEmbedded` as well)

lavalleeale commented on site/ts/ImageAnnotationEmbedded.ts:
```diff
@@ -0,0 +1,800 @@
+import type { AnnotationEditor } from '@markerjs/markerjs-ui';
+
+// Type declarations for external libraries
+const markerjs3 = window.markerjs3;
+const markerjsUI = window.markerjsUI;
+const $ = window.$;
```
We should remove these variables and use them from window directly

lavalleeale reviewed (changes requested):
Added code comments but functionality works well for me

martig7 committed:
Update site/app/controllers/pdf/PDFController.php

Co-authored-by: Alex Lavallee <alex@lavallee.one>

martig7 committed:
Update site/app/controllers/pdf/PDFController.php

Co-authored-by: Alex Lavallee <alex@lavallee.one>

martig7 committed:
Fix Jeffrey's bug

martig7 committed:
Fix lint, remove unnecessary declarations

williamschen23 commented on site/app/controllers/pdf/PDFController.php:
```diff
@@ -251,4 +302,151 @@ public function showGraderPDFEmbedded(string $gradeable_id) {
+        $is_anon = $_POST['is_anon'] ?? false;
+        $filename = html_entity_decode($filename);
+        $file_path = urldecode($_POST['file_path']);
+        $real_path = (bool) $is_anon ? "" : $file_path;
```
/**
     * Transforms non-boolean values to boolean values, i.e. 'true' to true
     * @param mixed $variable The variable that is being passed in. Type of mixed to allow for boolean, string, integer, or null.
     */
    public static function getBooleanValue(mixed $variable): bool {
    
    we have this

martig7 replied:
used filter_var, it's native to php and does exactly what I want

martig7 committed:
Fix linting, stop casting everything to bool

martig7 committed:
Merge branch 'main' into add-image-annotation

jeffrey-cordero reviewed (approved):
The work so far to implement image annotation looks solid, laying great groundwork for further optimizations and new features.

williamschen23 commented on site/app/templates/grading/electronic/ImageAnnotationEmbedded.twig:
```diff
@@ -0,0 +1,44 @@
+{% for stylesheet in localcss %}
+    <link rel="stylesheet" type="text/css" href="{{ stylesheet }}"/>
+{% endfor %}
+{# MarkerJS scripts are now loaded globally in the main TA grading page #}
+{% for script in localjs %}
+    <script src="{{ script }}"></script>
+{% endfor %}
```
why not use `addInternalJs` and `addInternalCss`?

williamschen23 reviewed (approved):
Code looks decent and functionality works as intended

williamjallen commented on site/app/controllers/pdf/PDFController.php:
```diff
@@ -251,4 +304,155 @@ public function showGraderPDFEmbedded(string $gradeable_id) {
+            }
+        }
+        else {
+            $anon_path = $this->getPath($file_path, $gradeable_id, false);
```
`getPath()` doesn't do any form of path sanitization when the third argument is false.  I believe that means that this code allows an arbitrary read if a carefully constructed `$_POST['file_path']` argument is provided.  It would be better to use `FileUtils::joinPaths()` in `getPath()` at a minimum, but it would be even better to just get rid of any sort of logic where we pass paths from the client side to the server side.  

Even if these security vulnerabilities are fixed now, passing paths around unnecessarily opens the door to developers unintentionally creating new security vulnerabilities in the future.

williamjallen replied:
The previous PDF annotator is unfortunately something we've had security issues with in the past, so it's best to not rely on it too much.  In general, I think passing the most minimal part needed to uniquely identify a file is the best way to approach that.  Is that just the filename, or a part of the filename?

NicholasCiuica replied:
I am a little confused, what downsides could using joinPaths() have in this situation?

williamjallen commented on site/app/controllers/pdf/PDFController.php:
```diff
@@ -251,4 +304,155 @@ public function showGraderPDFEmbedded(string $gradeable_id) {
+        $annotation_dir = FileUtils::joinPaths($this->core->getConfig()->getCoursePath(), 'annotations', $gradeable_id, $id, $active_version);
+        $annotation_jsons = [];
+        $file_path_md5 = md5($file_path);
+        if (is_dir($annotation_dir)) {
+            $dir_iter = new \FilesystemIterator($annotation_dir);
+            foreach ($dir_iter as $annotation_file) {
+                if (explode('_', $annotation_file->getFilename())[0] === $file_path_md5) {
+                    $file_contents = file_get_contents($annotation_file->getPathname());
+                    $annotation_decoded = json_decode($file_contents, true);
+                    if ($annotation_decoded !== null) {
+                        $grader_id = $annotation_decoded["grader_id"];
+                        $annotation_jsons[$grader_id] = json_encode($annotation_decoded['annotations']);
+                    }
+                }
+            }
+        }
```
Can you add comments or factor some of this out into a separate function so it's clear what's going on here?

williamjallen commented on site/app/controllers/pdf/PDFController.php:
```diff
@@ -77,6 +77,54 @@ public function showStudentPDF(string $gradeable_id, string $filename, string $p
+        $annotation_path = FileUtils::joinPaths($this->core->getConfig()->getCoursePath(), 'annotations', $gradeable_id, $id, $active_version);
+        $annotation_jsons = [];
+        $md5_path = md5($path);
+        if (is_dir($annotation_path)) {
+            $dir_iter = new \FilesystemIterator($annotation_path);
+            foreach ($dir_iter as $file_info) {
+                if (explode('_', $file_info->getFilename())[0] === $md5_path) {
+                    $file_contents = file_get_contents($file_info->getPathname());
+                    $annotation_decoded = json_decode($file_contents, true);
+                    if ($annotation_decoded !== null) {
+                        $grader_id = $annotation_decoded["grader_id"];
+                        $annotation_jsons[$grader_id] = json_encode($annotation_decoded['annotations']);
+                    }
+                }
+            }
+        }
```
Can any of this be shared with the similar chunk in `showGraderImageEmbedded()`?

williamjallen commented on site/app/controllers/MiscController.php:
```diff
@@ -174,7 +174,8 @@ public function displayFile($dir = null, $path = null, $gradeable_id = null, $us
             // Check access through Access library
-            if (!$this->core->getAccess()->canI("path.read", ["dir" => $dir, "path" => $path])) {
+            $owns_annotated_file = filter_var($view_annotated_file, FILTER_VALIDATE_BOOL) && $dir === 'submissions_processed' && strpos($path, $this->core->getUser()->getId()) !== false;
+            if (!$this->core->getAccess()->canI("path.read", ["dir" => $dir, "path" => $path]) && !$owns_annotated_file) {
```
Wouldn't it be better to add this logic to `path.read`?

martig7 replied:
Done

williamjallen commented on site/app/views/ImageView.php:
```diff
@@ -0,0 +1,92 @@
+        $display_file_url = $this->core->buildCourseUrl(['display_file']);
+
+        $localcss = [];
+        $localcss[] = $this->core->getOutput()->timestampResource(FileUtils::joinPaths('image', 'image_annotation.css'), 'css');
```
You could use `addInternalCss()` here instead.

williamjallen commented on site/app/views/ImageView.php:
```diff
@@ -0,0 +1,92 @@
+        $localjs = [];
+
+        if ($jquery) {
+            $localjs[] = $this->core->getOutput()->timestampResource(FileUtils::joinPaths('jquery', 'jquery.min.js'), 'vendor');
```
You could use `addVendorJs()` here.

NicholasCiuica replied:
Could you explain how this code segment would be replaced by the use of addVendorJs()? I am unfamiliar with vendor and am unsure what the original code is doing.

williamjallen commented on site/app/views/ImageView.php:
```diff
@@ -0,0 +1,92 @@
+                'download_path' => $download_path
+            ]);
+        }
+        catch (\Exception $e) {
+            $this->core->addErrorMessage("Error rendering twig template: " . $e->getMessage());
+            $this->core->addErrorMessage("Stack trace: " . $e->getTraceAsString());
+            $this->core->getOutput()->renderJsonError($e->getMessage());
+        }
```
Let's just let any exceptions here fall through so they show up in the logs.

williamjallen commented on site/globals.d.ts:
```diff
@@ -1,5 +1,7 @@
 import * as luxon from 'luxon';
 import * as Twig from 'twig';
+import * as markerjs3 from '@markerjs/markerjs3';
+import * as markerjsUI from '@markerjs/markerjs-ui';
```
How big are these libraries?

NicholasCiuica replied:
From my testing it looks like the markerjs libraries are only loaded on the grading page where they are used.

martig7 committed:
Refactor annotation json loader out

martig7 committed:
Add annotation ownership check to access library

martig7 committed:
Remove unnecessary comment

martig7 committed:
Remove error checking so that they show in logs

martig7 committed:
Remove Whitespace

dagemcn committed:
Merge branch 'main' into add-image-annotation

NicholasCiuica committed:
Merge branch 'main' into add-image-annotation

NicholasCiuica committed:
change not saved color

"not saved" status is now a different color from the "saved" status

NicholasCiuica committed:
php linting fixes

NicholasCiuica committed:
Merge branch 'main' into add-image-annotation

NicholasCiuica committed:
another php linting fix

NicholasCiuica committed:
Merge branch 'add-image-annotation' of https://github.com/Submitty/Submitty into add-image-annotation

NicholasCiuica committed:
Revert "another php linting fix"

This reverts commit 4f6ee3b8affa800696528c676369c9965ccbcf3a.

NicholasCiuica committed:
actual php lint fix

NicholasCiuica committed:
this should be the last php linting fix

NicholasCiuica committed:
title for annotation buttons

NicholasCiuica committed:
CSS changes part 1

fixing css linting issues while also ensuring high contrast on buttons

NicholasCiuica committed:
css update part 2

making image error look like danger alert

NicholasCiuica committed:
Merge branch 'main' into add-image-annotation

NicholasCiuica committed:
fixing up pseudoclasses for colors

NicholasCiuica committed:
Merge branch 'add-image-annotation' of https://github.com/Submitty/Submitty into add-image-annotation

NicholasCiuica committed:
another css fix

NicholasCiuica committed:
moving around :active:focus

NicholasCiuica committed:
Update image_annotation.css

NicholasCiuica committed:
reduced editor wrapper func code duplication

put all logic for creating and hiding editor wrappers to reduce code duplication. this will also make it easier to keep track of annotation page state (open/closed)

NicholasCiuica committed:
importing editor open check

NicholasCiuica committed:
moving keybind disable logic to ta-grading-keymap

NicholasCiuica committed:
Merge branch 'main' into add-image-annotation

NicholasCiuica committed:
img_annotation_paths now stores all annotations

NicholasCiuica commented:
After testing this PR and making some fixes to get it merge-ready, I found a missing feature that constitutes its own issue/PR after this one is merged. In this PR, students will only see at most one annotated image per image file they submit, even if multiple graders make annotations to the same file. This causes problems when multiple graders try to annotate the same problem, or two graders are annotating different problems on the same page.

Leaving this image to find the issue again easily. All the work needed to fix this issue will probably be in \site\ts\ImageAnnotationEmbedded.ts.
<img width="562" height="209" alt="Screenshot 2026-06-29 141343" src="https://github.com/user-attachments/assets/18a1794b-c773-4100-b10b-51e3882c518c" />

NicholasCiuica committed:
Revert "img_annotation_paths now stores all annotations"

This reverts commit 685d544a5d903e34cc43eb00aac1e2af8986ae36.

NicholasCiuica committed:
Merge branch 'main' into add-image-annotation

NicholasCiuica commented on site/app/controllers/pdf/PDFController.php:
```diff
@@ -77,6 +77,54 @@ public function showStudentPDF(string $gradeable_id, string $filename, string $p
+
+        $annotation_path = FileUtils::joinPaths($this->core->getConfig()->getCoursePath(), 'annotations', $gradeable_id, $id, $active_version);
+        $annotation_jsons = [];
+        $md5_path = md5($path);
```
Is it an important change that we find a better hashing function for this? If not, this change would be better explored later, and we should prioritize getting this PR ready for merge.

NicholasCiuica committed:
moving media css to bottom

NicholasCiuica committed:
experimental change adding css and js differently

this current change does not work. I tried to use addInternalCss and addVendorJs here could not yet get them to work

NicholasCiuica committed:
removing unnecessary jquery load

jquery is already installed globally

NicholasCiuica committed:
removing commented-out code

NicholasCiuica committed:
removed unreached getPath

the getPath only ran if this route was invoked on a non-anonymous file path, but the only use of this route is on an anonymous file path, so this line is never reached. Furthermore, if it was reached, then the getPath could cause an arbitrary read because it does not perform sanitation on the input file_path when its third parameter is false.

NicholasCiuica committed:
Merge branch 'main' into add-image-annotation

NicholasCiuica committed:
css fix

NicholasCiuica committed:
linting fixes

NicholasCiuica committed:
css lint fix

NicholasCiuica committed:
eslint issue?

it seems that eslint is not picking up on a number of issues that are being picked up locally

NicholasCiuica committed:
Revert "eslint issue?"

This reverts commit fc31b89c06b8c8c43880f426c29f08103d72f12e.

NicholasCiuica committed:
moved editor css out of ts

NicholasCiuica committed:
Merge branch 'main' into add-image-annotation

NicholasCiuica committed:
css linting

NicholasCiuica committed:
Merge branch 'add-image-annotation' of https://github.com/Submitty/Submitty into add-image-annotation

NicholasCiuica committed:
moving css import next to markerjs import

for clarity

mattLif reviewed (approved):
Code looks good after reviewing. Tested feature with an uploaded image and observed the expected behavior. Just be sure to mention in documentation that annotations are only sent to student once rubric is fully graded. Also noticed that there's conflicts with multiple annotations, but that is being fixed in a future PR. I'm approving.

NicholasCiuica committed:
Merge branch 'main' into add-image-annotation
